rfcs/proposed/task_group_dynamic_dependencies/README.md

```diff
@@ -0,0 +1,369 @@
+are complete.
+3. **Executing:** has started executing its body but is not yet complete.
+4. **Completed:** fully executed to completion.
+
+In the current `task_group` specification, the `task_group::defer` function
+already allows separate task creation and submission.
+`task_group::defer` returns a `tbb::task_handle` that represents a created
+task. A created task remains created until it is submitted through
+`task_group::run` or `task_group::run_and_wait`. The current
+`task_group` specification treats accessing a `task_handle` after it is submitted
+via one of the run functions as undefined behavior. Therefore, a
+`task_handle` represents a created task only. Furthermore, since `task_group`
+does not support task dependencies, any task that is run can be immediately
+scheduled for execution without considering dependencies.
+
+The first extension is to expand the semantics and usable lifetime of
+`task_handle` so that it can used as a predecessor to other tasks even after it is
+passed to run. The handle will track the state of its associated task and may represent
+a task that is created, submitted, executing, or completed. Similarly, a submitted `task_handle`
+may represent a task that depends on predecessors that must complete before it can execute.
+In that case, passing a `task_handle` to `task_group::run` or `task_group::run_and_wait` only makes
+it available for dependency tracking but does not make it immediately eligible for execution.
+
+Since `task_group::run` and `task_group::run_and_wait` currently accept `task_handle` as an rvalue that
+allows the `task_group` to move-construct or move-assign from the received handle and transit it into the
+empty state. Hence, some changes in `run` and `run_and_wait` semantics would also be required to handle the
+`task_handle` argument differently to support the extension of its useful lifetime. One of the possible options
+is to add overloads for `run` and `run_and_wait` accepting `const task_handle&` argument.
+
+An other aspect that should be addressed is co-existence of the proposed semantics for `task_handle` with the currently
+available preview feature that allows to return `task_handle` from the function object passed into `run` or `run_and_wait`. 
+The option is to allow also returning the `task_handle` representing a task with added dependencies from the function object
+to make it visible for tracking the dependencies.
+
+### Add Functions to Set Dependencies.
+
+The next logical extension is to add a mechanism for specifying dependencies
+between tasks. In the most conservative view, it should only be allowed to add
+predecessors (in-dependencies) to tasks in the created state.
+After a task starts, adding more predecessors is irrelevant,
+since it’s too late to delay the task execution.
+
+It might be reasonable to add new predecessors to a task that is
+currently executing if it is suspended until these additional dependencies complete.
```

Comment:
Alternatively, the "suspension support" can be added via the `tg.wait_for(task_handle)` future extension.